infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -0,0 +1,513 @@
+    private CorfuRuntime getCorfuRuntime() {
+        if (runtime == null) {
+            localCorfuEndpoint = getCorfuEndpoint(localHost, localClusterDescriptor.getCorfuPort());
+            log.debug("Connecting to local Corfu {}", localCorfuEndpoint);
+            runtime = CorfuRuntime.fromParameters(CorfuRuntime.CorfuRuntimeParameters.builder()
+                    .trustStore((String) serverContext.getServerConfig().get("--truststore"))
+                    .tsPasswordFile((String) serverContext.getServerConfig().get("--truststore-password-file"))
+                    .keyStore((String) serverContext.getServerConfig().get("--keystore"))
+                    .ksPasswordFile((String) serverContext.getServerConfig().get("--keystore-password-file"))
+                    .tlsEnabled((Boolean) serverContext.getServerConfig().get("--enable-tls"))
+                    .build())
+                    .parseConfigurationString(localCorfuEndpoint).connect();
+        }
+
+        return runtime;
+    }
+
+    /**
+     * Verify current node belongs to a cluster in the topology.
+     */
+    private boolean clusterPresentInTopology(TopologyDescriptor topology) {
+        localClusterDescriptor = topology.getClusterDescriptor(localEndpoint);
+        if (localClusterDescriptor != null) {
+            localNodeDescriptor = localClusterDescriptor.getNode(localEndpoint);
+        }
+
+        return localClusterDescriptor != null && localNodeDescriptor != null;
+    }
+
+    /**
+     * Retrieve local Corfu Endpoint
+     */
+    private String getCorfuEndpoint(String localEndpoint, int corfuPort) {
+        return NodeLocator.parseString(localEndpoint).getHost() + ":" + corfuPort;
+    }
+
+    /**
+     * Retrieve Log Replication Configuration.
+     *
+     * This configuration represents all common parameters for the log replication, regardless of
+     * a cluster's role.
+     */
+    private LogReplicationConfig getLogReplicationConfiguration(CorfuRuntime runtime) {
+
+        LogReplicationStreamNameTableManager replicationStreamNameTableManager =
+                new LogReplicationStreamNameTableManager(runtime, pluginFilePath);
+
+        Set<String> streamsToReplicate = replicationStreamNameTableManager.getStreamsToReplicate();
+
+        // TODO pankti: Check if version does not match.  If if does not, create an event for site discovery to
+        //  do a snapshot sync.
+        // TODO(Gabriela): pending review upgrade path (changes)
+        boolean upgraded = replicationStreamNameTableManager
+                .isUpgraded();
+
+        if (upgraded) {
+            input(new DiscoveryServiceEvent(DiscoveryServiceEvent.DiscoveryServiceEventType.UPGRADE));
+        }
+
+        return new LogReplicationConfig(streamsToReplicate);
+    }
+
+    /**
+     * Register interest on Log Replication Lock.
+     *
+     * The node that acquires the lock will drive/lead log replication.
+     */
+    private void registerToLogReplicationLock() {
```

Comment:
Right  now this will continue indefinitely in case of constant failures, why do we never want  to timeout?

Reply:
We can discuss this, currently, we just need a node to be leader... otherwise no workflow is triggered. So somehow it's a requirement. But yes, we can also have a number of retries or a timeout and completely abort if we fail to register to the lock...